src/main/java/edu/ucla/library/prl/harvester/Institution.java

```diff
@@ -0,0 +1,253 @@
+        myName = Objects.requireNonNull(aName);
+        myDescription = Objects.requireNonNull(aDescription);
+        myLocation = Objects.requireNonNull(aLocation);
+
+        if (anEmail == null && aPhone == null && aWebContact == null) {
+            throw new NullPointerException(LOGGER.getMessage(MessageCodes.PRL_002));
+        } else {
+            myEmail = Optional.ofNullable(anEmail);
+            myPhone = Optional.ofNullable(aPhone);
+            myWebContact = Optional.ofNullable(aWebContact);
+        }
+
+        myWebsite = Objects.requireNonNull(aWebsite);
+    }
+
+    /**
+     * Instantiates an institution from its JSON representation.
+     * <p>
+     * <b>This constructor is meant to be used only by generated service proxy code!</b>
+     * {@link #Institution(String, String, String, InternetAddress, PhoneNumber, URL, URL)} should be used everywhere
+     * else.
+     *
+     * @param aJsonObject An institution represented as JSON
+     * @throws IllegalArgumentException If the JSON representation is invalid
+     */
+    @SuppressWarnings({ "PMD.AvoidCatchingNPE", "PMD.AvoidCatchingGenericException" })
+    public Institution(final JsonObject aJsonObject) {
+        try {
+            myName = Objects.requireNonNull(aJsonObject.getString(NAME));
+            myDescription = Objects.requireNonNull(aJsonObject.getString(DESCRIPTION));
+            myLocation = Objects.requireNonNull(aJsonObject.getString(LOCATION));
+            myEmail = Optional.ofNullable(aJsonObject.getString(EMAIL)).map(email -> {
+                try {
+                    return new InternetAddress(email, true);
+                } catch (final AddressException details) {
+                    throw new IllegalArgumentException(details.getMessage(), details);
+                }
+            });
+            myPhone = Optional.ofNullable(aJsonObject.getString(PHONE)).map(phone -> {
+                try {
+                    return PHONE_NUMBER_UTIL.parse(phone, null);
+                } catch (NumberParseException details) {
+                    throw new IllegalArgumentException(details.getMessage(), details);
+                }
+            });
+            myWebContact = Optional.ofNullable(aJsonObject.getString(WEB_CONTACT)).map(webContact -> {
+                try {
+                    return new URL(webContact);
+                } catch (MalformedURLException details) {
+                    throw new IllegalArgumentException(details.getMessage(), details);
+                }
+            });
+
+            if (myEmail.isEmpty() && myPhone.isEmpty() && myWebContact.isEmpty()) {
+                throw new IllegalArgumentException(LOGGER.getMessage(MessageCodes.PRL_002));
+            }
+
+            myWebsite = new URL(Objects.requireNonNull(aJsonObject.getString(WEBSITE)));
+        } catch (final MalformedURLException | NullPointerException details) {
```

Comment:
I think it's fuzzy, but an empty JsonObject (or even one that is missing fields) is not "no-argument" IMO (which would be if `null` itself is passed to the constructor). The no-argument case isn't handled here, but probably should be. What if we added the following statement to the top of the method (and same for that of `Job`)?

```java
Objects.requireNonNull(aJsonObject);
```

Also:

> it seems the PMD.AvoidCatchingNPE and PMD.AvoidCatchingGenericException exceptions are used to work around wanting to use IllegalArgumentException

The former is actually to get around having to do this for every field (by using `Objects#requireNonNull` instead):

```java
final String name = aJsonObject.getString(NAME);

if (name == null) {
    throw new IllegalArgumentException("Missing 'name' property"); // The "argument" being aJsonObject
} else {
    myWebsite = new URL(name);
}
```

The latter is because of the compilation errors I mentioned in the description.

Reply:
Oh, okay on `Objects.requireNonNull(aJsonObject);` I wasn't considering that was for the JsonObject itself. I thought it was for the URL constructor. Yes, I agree this would be good to add.

Reply:
Just want to point out that the NPE's caused by calling `Objects#requireNonNull` on the field getters are being marked as the cause of the resulting llegalArgumentExceptions.

You don't think there should be a distinction made in the type of exception thrown for `aJsonObject` being `null` vs. `aJsonObject` maybe having all required fields except for, say, "website"?

Reply:
Yes, I think the aJsonObject being null should be a NullPointerException and that the aJsonObject missing required fields should be InvalidInstitutionException. I wasn't considering the aJsonObject being null, but you're right it would be nice to check for that. Fwiw, that requireNonNull method can also take a string that gives more detail to the NPE.

Reply:
Right, my bad. What is it that is inadequate about IllegalArgumentException that InvalidInstitutionException would remedy?

Reply:
My quibble isn't with it as a wrapper for NumberFormatException or MalformedUrlException. In those cases, I think it's appropriate, but I'm not happy with it as a wrapper for "One of myEmail, myPhone, or myWebContact is missing" or "I get a null for `aJsonObject.getString(WEBSITE)`". Something missing isn't an IllegalArgumentException, to me, as much as it is "the object data I was passed doesn't contain all the components that are required to create this Institution object" (whatever you call that more specific exception).

Reply:
Okay, I think the "more descriptive" argument is strong in favor of using a custom exception vs. IllegalArgumentException. It has to be an unchecked exception though, since generated service proxy code unfortunately can't catch any checked exceptions.

Will work on making that change.

Reply:
Pushed